core/trie/trie.go

```diff
@@ -11,11 +11,70 @@ import (
 	"github.com/NethermindEth/juno/core/crypto"
 	"github.com/NethermindEth/juno/core/felt"
 	"github.com/NethermindEth/juno/db"
+	"github.com/NethermindEth/juno/db/memory"
 	"github.com/NethermindEth/juno/utils"
 )
 
 const globalTrieHeight = 251 // TODO(weiihann): this is declared in core also, should be moved to a common place
 
+// TODO(maksym): try to make this interface private after common trie is integrated
+type TrieReader struct {
+	height uint8
+	// TODO(maksym): Trie mutates rootKey field, and the idea behind TrieReader is to be immutable.
+	// Fix this in the Trie logic (potentially modify the rootKeyIsDirty field)
+	rootKey     *BitArray
+	maxKey      *felt.Felt
+	readStorage *ReadStorage
+	hash        crypto.HashFn
+}
+
+type NewTrieReaderFunc func(db.KeyValueReader, []byte, uint8) (*TrieReader, error)
+
+func NewTrieReaderPedersen(
+	r db.KeyValueReader,
+	prefix []byte,
+	height uint8,
+) (*TrieReader, error) {
+	return newTrieReader(r, prefix, height, crypto.Pedersen)
+}
+
+func NewTrieReaderPoseidon(
+	r db.KeyValueReader,
+	prefix []byte,
+	height uint8,
+) (*TrieReader, error) {
+	return newTrieReader(r, prefix, height, crypto.Poseidon)
+}
+
+func newTrieReader(
+	r db.KeyValueReader,
+	prefix []byte,
+	height uint8,
+	hash crypto.HashFn,
+) (*TrieReader, error) {
+	if height > felt.Bits {
+		return nil, fmt.Errorf("max trie height is %d, got: %d", felt.Bits, height)
+	}
+
+	// maxKey is 2^height - 1
+	maxKey := new(felt.Felt).Exp(new(felt.Felt).SetUint64(2), new(big.Int).SetUint64(uint64(height)))
+	maxKey.Sub(maxKey, new(felt.Felt).SetUint64(1))
+
+	readStorage := NewReadStorage(r, prefix)
+	rootKey, err := readStorage.RootKey()
+	if err != nil && !errors.Is(err, db.ErrKeyNotFound) {
+		return nil, err
+	}
+
+	return &TrieReader{
+		readStorage: readStorage,
+		height:      height,
+		rootKey:     rootKey,
+		maxKey:      maxKey,
+		hash:        hash,
+	}, nil
+}
```

Comment:
Should this type be returned by reference? The answer is yes if:
- it's shared by different go-routines
- it has concurrency control mechanisms (such as a mutex) 

Otherwise, I think the answer is maybe. Let me see how it is used

Reply:
Its not used by different goroutines. It's not a big structure as well, so we may return it by value